homework/ba-nguyen/lesson-7/longest-zigzag-path-in-a-binary-tree.ts

```diff
@@ -0,0 +1,53 @@
+class TreeNode {
+  val: number;
+  left: TreeNode | null;
+  right: TreeNode | null;
+  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
+    this.val = val === undefined ? 0 : val;
+    this.left = left === undefined ? null : left;
+    this.right = right === undefined ? null : right;
+  }
+}
+
+function longestZigZag(root: TreeNode | null): number {
+  let res = 0;
+
+  function dfs(root: TreeNode | null) {
+    if (!root) return [-1, -1];
+
+    const left = dfs(root.left);
+    const right = dfs(root.right);
+    const pathLeft = left[1] + 1;
+    const pathRight = right[0] + 1;
+
+    res = Math.max(res, pathLeft || res, pathRight);
+
+    return [pathLeft, pathRight];
+  }
+  dfs(root);
+
+  return res;
+}
+
+function longestZigZag1(root: TreeNode | null): number {
+  let res = 0;
+
+  function dfs(root: TreeNode | null, direction: string, step: number) {
```

Comment:
curious: in ts, when you pass a string to a function, will a copy of string be created?

Reply:
String is immutable so whenever you try to change it. But Object and Array in js is opposite, these are reference

Reply:
does that mean every time you call dfs, a new string is created?

Reply:
yes, exactly. You mean like it affect the performance of my code, don't you?

Reply:
Ohh i understand your thinking
Ex:
function changeChar(char) { char += '2' }
let char = '1'
changeChar(char) // => result: '12'

Is that your thinking?

Reply:
oh nvm, I think cannot apply in this situation.

Reply:
In javascript as well as typescript, only Object and Array is reference
Ex: 
const obj = { x: 1 }
function changeValueOfObj(obj) { obj.x = 2 }
changeValueOfObj(obj) // result: obj = { x: 2 }